tools/tracing/timeline/EXTENSION.md

```diff
@@ -0,0 +1,196 @@
+# Extending the timeline tool
+
+This document is mainly for VM binding developers who want to extend this timeline tool to trace and
+visualize VM-specific events on the timeline.
+
+## Custom work packets in VM bindings
+
+mmtk-core contains trace points that captures the beginning and end of *all* work packets.  If a VM
+bindings defines its own work packets and execute them, they will automatically appear on the
+timeline, without needing to modify or extend any scripts.
+
+But if you wish to add additional attributes to work packets or events and browse them in Perfetto
+UI, please read on.
+
+## The output format of `capture.bt`
+
+The capturing script `capture.bt` prints events in a text format.  Each line contains
+comma-separated values:
+
+```
+name,ph,tid,timestamp,arg0,arg1,arg2,...
+```
+
+The `visualize.py` script will transform those lines into the [Trace Event Format], a JSON-based
+format suitable for Perfetto UI, like this: `{"name": name, "ph": ph, "tid": tid, "ts": ts}`.
+Possible values of the event type (or "phase", "ph") are defined by the [Trace Event Format].  For
+example, "B" and "E" represent the beginning and the end of a duration event, while "i" represents
+an instant event.  Additional arguments (arg0, arg1, ...) are processed by `visualize.py` in ways
+specific to each event.  Some arguments are added to the resulting JSON object, for example
+`{"name": name, ..., "args": {"is_roots": 1, "num_slots": 2}}`  The data in "args" are
+human-readable, and can be displayed on Perfetto UI.
+
+[Trace Event Format]: https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU
+
+## Extending the capturing script
+
+VM binding developers may insert custom USDT trace points into the VM binding.  They need to be
```

Comment:
Let's mention that the custom probes can reference the global variables defined in `capture.bt`, such as `@harness`, etc.

Reply:
Yes.  That makes sense.  Similarly, it is intentional that the extension Python script for `visualize.py` can access the `LogProcessor` instance.  I'll mention that, too.

Reply:
I added comments about variables accessible by the custom probes.  I also added an "advanced usage" section to mention the "log_processor" parameter for the extension script of visualize.py.